curve_parameters/new_curve_script.py

```diff
@@ -33,157 +15,251 @@ def to_hex(val, length):
-    s += "   static constexpr storage<limbs_count> modulus_4 = {"+to_hex(modulus*4,8*limbs)[:-2]+"};\n"
-    s += "   static constexpr storage<2*limbs_count> modulus_wide = {"+to_hex(modulus,8*limbs*2)[:-2]+"};\n"
-    s += "   static constexpr storage<2*limbs_count> modulus_sqared = {"+to_hex(modulus*modulus,8*limbs)[:-2]+"};\n"  
-    s += "   static constexpr storage<2*limbs_count> modulus_sqared_2 = {"+to_hex(modulus*modulus*2,8*limbs)[:-2]+"};\n"   
-    s += "   static constexpr storage<2*limbs_count> modulus_sqared_4 = {"+to_hex(modulus*modulus*2*2,8*limbs)[:-2]+"};\n"   
-    s += "   static constexpr unsigned modulus_bits_count = "+str(modulus_bits_count)+";\n"
-    m = int(math.floor(int(pow(2,2*modulus_bits_count) // modulus)))
-    s += "   static constexpr storage<limbs_count> m = {"+ to_hex(m,8*limbs)[:-2] +"};\n"
-    s += "   static constexpr storage<limbs_count> one = {"+ to_hex(1,8*limbs)[:-2] +"};\n"
-    s += "   static constexpr storage<limbs_count> zero = {"+ to_hex(0,8*limbs)[:-2] +"};\n"
-
-    if ntt:
-        for k in range(size):
-            omega = get_root_of_unity(int(pow(2,k+1)))
-            s += "   static constexpr storage<limbs_count> omega"+str(k+1)+"= {"+ to_hex(omega,8*limbs)[:-2]+"};\n"
-        s += omega
-        for k in range(size):
-            omega = get_root_of_unity(int(pow(2,k+1)))
-            s += "   static constexpr storage<limbs_count> omega_inv"+str(k+1)+"= {"+ to_hex(pow(omega, -1, modulus),8*limbs)[:-2]+"};\n"
-        s += omega_inv
-        for k in range(size):
-            s += "   static constexpr storage<limbs_count> inv"+str(k+1)+"= {"+ to_hex(pow(int(pow(2,k+1)), -1, modulus),8*limbs)[:-2]+"};\n"  
-        s += inv
-    s+=" };\n"   
-    return s
-
-def create_gen():
-    s = " struct group_generator {\n"
-    s += "  static constexpr storage<fq_config::limbs_count> generator_x = {"+to_hex(gen_x,8*limb_q)[:-2]+ "};\n"
-    s += "  static constexpr storage<fq_config::limbs_count> generator_y = {"+to_hex(gen_y,8*limb_q)[:-2]+ "};\n"
-    s+=" };\n" 
-    return s
-
-def get_config_file_content(modolus_p, bit_count_p, limb_p, ntt_size, modolus_q, bit_count_q, limb_q, weierstrass_b):
-    file_content = ""
-    file_content += "#pragma once\n#include \"../../utils/storage.cuh\"\n"
-    file_content += "namespace PARAMS_"+curve_name.upper()+"{\n"
-    file_content += create_field_parameters_struct(modolus_p,bit_count_p,limb_p,True,ntt_size,"fp_config")
-    file_content += create_field_parameters_struct(modolus_q,bit_count_q,limb_q,False,0,"fq_config")
-    file_content += " static constexpr unsigned weierstrass_b = " + str(weierstrass_b)+ ";\n"
-    file_content += create_gen()
-    file_content+="}\n"
-    return file_content
+        s += f'0x{c}, '
+        
+    return s[:-2]
+
+
+def get_root_of_unity(modulus_p, order: int) -> int:
+    assert (modulus_p - 1) % order == 0
+    return pow(5, (modulus_p - 1) // order, modulus_p)
```

Comment:
why 5?

can u use a roo_of_unity?

this is what my json looks like for bn254:

```
{
    "curve_name" : "bn254",
    "modolus_p" : 21888242871839275222246405745257275088548364400416034343698204186575808495617,
    "bit_count_p" : 254,
    "limb_p" :  8,
    "ntt_size" : 16,
    "modolus_q" : 21888242871839275222246405745257275088696311157297823662689037894645226208583,
    "bit_count_q" : 254,
    "limb_q" : 8,
    "weierstrass_b" : 3,
    "gen_x" : 1,
    "gen_y" : 2,
    "root_of_unity": 421743594562400382753388642386256516545992082196004333756405989743524594615
}
```

I then did:

```
def get_root_of_unity(order: int) -> int:
    assert (modolus_p - 1) % order == 0
    return pow(root_of_unity, (1 << ntt_size) // order, modolus_p)
```

Reply:
@ImmanuelSegol I'm not sure.. the only change on these lines is whitespace.

@guy-ingo Whats the reason behind a hard coded 5 here?

Reply:
@jeremyfelder yeah 5 being hard-coded there is not your change. But IMO we do need to use root of unity and not just 5 because first of all, 5 might not be the primitive root of the multiplicative subgroup of some fields, which would mean that it generates invalid twiddle factors. And secondly, we will want to use the exact same twiddle factors that the applications we want to support use, so being able to choose here is important